controller/controller.go

```diff
@@ -1,81 +1,248 @@
+	for {
+		select {
+		case row, ok := <-writeChan:
+			if !ok {
+				return
+			}
+			readChan <- row
+		}
+	}
+}
+
+func WorkerPoolHandler(w http.ResponseWriter, r *http.Request) {
+	numberType := strings.ToLower(r.URL.Query().Get("type"))
+	if (numberType != "odd" && numberType != "even") || numberType == "" {
+		errorResponse := model.ErrorResponse{Err: "Invalid or empty param Type."}
+		render.New().JSON(w, http.StatusBadRequest, errorResponse)
+		return
+	}
+
+	items, err := strconv.Atoi(r.URL.Query().Get("items"))
+	if err != nil || items == 0 {
+		log.Println("Error converting items param to int.")
+		errorResponse := model.ErrorResponse{Err: "Invalid or empty items param."}
+		render.New().JSON(w, http.StatusBadRequest, errorResponse)
+		return
+	}
+
+	itemsPerWorker, err := strconv.Atoi(r.URL.Query().Get("items_per_workers"))
+	if err != nil || itemsPerWorker == 0 {
+		log.Println("Error converting items_per_workers param to int.")
+		errorResponse := model.ErrorResponse{Err: "Invalid or empty items per worker param."}
+		render.New().JSON(w, http.StatusBadRequest, errorResponse)
+		return
+	}
+
+	if itemsPerWorker > items {
+		errorResponse := model.ErrorResponse{Err: "Items per worker param should be lower than items param."}
+		render.New().JSON(w, http.StatusBadRequest, errorResponse)
+	}
+
+	numberOfWorkers := items / itemsPerWorker
+
+	csvfile, err := os.Open(config.FirstGenCSVPath)
+	if err != nil {
+		log.Fatal(err)
 	}
+	defer csvfile.Close()
+
+	reader := csv.NewReader(csvfile)
 
-	if wantedIndex < 1 || wantedIndex > 151 {
-		log.Println("Please introduce a valid pokemon ID from first gen. (1-151)")
-		return 0, errors.New("invalid id")
+	// create the input/output channels
+	srcCh := make(chan []string)
+	outputCh := make(chan []string, items)
+
+	// manage synchronization
+	var wg sync.WaitGroup
+
+	// declare the workers
+	for workerID := 1; workerID <= numberOfWorkers; workerID++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			worker(outputCh, srcCh)
+		}()
 	}
 
-	return wantedIndex, nil
+	rowNumber := 0
+	itemsToGet := items * 2 //total items to get, after odd or even choice
+	go func() {
+		for {
+			record, err := reader.Read()
+
+			if err == io.EOF {
+				fmt.Println("ending because of EOF")
+				break
+			} else if err != nil {
+				log.Println(err)
+				break
+			}
+			rowNumber += 1
+			if rowNumber <= itemsToGet {
+				idNum, err := strconv.Atoi(record[0])
+				if err != nil {
+					log.Println(err)
+					break
+				}
+				if numberType == "even" && (idNum%2 == 0) {
+					srcCh <- record
+				} else if numberType == "odd" && (idNum%2 == 1) {
+					srcCh <- record
+				}
+			} else {
+				break
+			}
+		}
+		close(srcCh)
+	}()
+
+	// wait for worker group to finish.
+	go func() {
+		wg.Wait()
+		close(outputCh)
+	}()
+
+	// format the response.
+	response := []model.PokemonData{}
+	for pokemonInfo := range outputCh {
+		pokemonID, err := strconv.Atoi(pokemonInfo[0])
```

Comment:
It would be better if the worker function did all of this processing, and you just read the converted information from the channel. That means your output channel would handle model.PokemonData